File: src/main/java/io/github/takenoha/towerdefense/runtime/TerrainMutationInput.java
package io.github.takenoha.towerdefense.runtime;

import java.util.Objects;

/** Paper-independent facts used to authorize one event-enemy block action. */
public record TerrainMutationInput(
        String currentMaterialKey,
        boolean currentInventoryHolder,
        boolean currentCore,
        String targetMaterialKey) {
    public TerrainMutationInput {
        currentMaterialKey = requireMaterialKey(currentMaterialKey, "currentMaterialKey");
        targetMaterialKey = requireMaterialKey(targetMaterialKey, "targetMaterialKey");
    }

    private static String requireMaterialKey(String value, String name) {
        Objects.requireNonNull(value, name);
        if (value.isBlank()) {
            throw new IllegalArgumentException(name + " must not be blank");
        }
        return value;
    }
}

File: src/main/kotlin/io/github/takenoha/towerdefense/runtime/TerrainMutationInput.kt
package io.github.takenoha.towerdefense.runtime

import java.util.Objects
import kotlin.jvm.JvmRecord

/** Paper-independent facts used to authorize one event-enemy block action. */
@JvmRecord
data class TerrainMutationInput(
    val currentMaterialKey: String,
    val currentInventoryHolder: Boolean,
    val currentCore: Boolean,
    val currentTileState: Boolean,
    val targetMaterialKey: String,
) {
    /** Keeps the original four-fact constructor source-compatible for non-tile callers. */
    constructor(
        currentMaterialKey: String,
        currentInventoryHolder: Boolean,
        currentCore: Boolean,
        targetMaterialKey: String,
    ) : this(currentMaterialKey, currentInventoryHolder, currentCore, false, targetMaterialKey)

    init {
        requireMaterialKey(currentMaterialKey, "currentMaterialKey")
        requireMaterialKey(targetMaterialKey, "targetMaterialKey")
    }

    companion object {
        private fun requireMaterialKey(value: String, name: String): String {
            Objects.requireNonNull(value, name)
            if (value.isBlank()) {
                throw IllegalArgumentException("$name must not be blank")
            }
            return value
        }
    }
}
